File: apps/web/lib/api/playback.ts
import { request } from "@/lib/http";
import type { MediaType } from "@/lib/media-types";

interface ApiEnvelope<T> {
  success: boolean;
  code: string;
  message: string;
  data: T;
}

export interface RecentWatchItem {
  media_item_id: number;
  library_id: number;
  kind: MediaType;
  title: string;
  year: number | null;
  poster_url: string | null;
  /** 电影横向背景剧照；缺失时前端用竖版海报生成模糊铺底。 */
  backdrop_url: string | null;
  /** 剧集最近播放那一集的 16:9 剧照；电影恒为 null。 */
  episode_still_url: string | null;
  season_number: number;
  episode_number: number;
  episode_title: string | null;
  /** 最近一次播放之后，同一媒体库中新进入且仍在位的分集数；电影恒为 0。 */
  new_episode_count: number;
  position_ms: number;
  duration_ms: number | null;
  progress_percent: number | null;
  played: boolean;
  play_count: number;
  last_played_at: string;
}

/** 当前账号在可见媒体库中的最近观看作品。 */
export async function listRecentWatch(limit = 20): Promise<RecentWatchItem[]> {
  const response = await request<ApiEnvelope<{ items: RecentWatchItem[] }>>(
    `/playback/recent?limit=${limit}`,
  );
  return response.data.items;
}

File: src/movieclaw_api/schemas/playback.py
"""播放记录在 Web 业务界面的响应模型。"""

from __future__ import annotations

from datetime import datetime

from movieclaw_api.schemas.base import BaseModel
from movieclaw_media.models import MediaKind


class RecentWatchItemView(BaseModel):
    """媒体库首页的一张最近观看卡片。"""

    media_item_id: int
    library_id: int
    kind: MediaKind
    title: str
    year: int | None
    poster_url: str | None
    backdrop_url: str | None
    episode_still_url: str | None
    season_number: int
    episode_number: int
    episode_title: str | None
    # 锚点之后、当前成员从未看过且文件在位的分集数——“还能接着看几集”，
    # 不是“最近入库了几集”：看完全剧、补齐旧季与洗版都不该触发提醒。
    unwatched_ahead_count: int
    position_ms: int
    duration_ms: int | None
    progress_percent: int | None
    played: bool
    play_count: int
    last_played_at: datetime


class RecentWatchView(BaseModel):
    """最近观看横排的数据载荷。"""

    items: list[RecentWatchItemView]
